### PR TITLE
intel: ssp: drain RX fifo when starting

### PR DESCRIPTION
**Pull request by kv2019i**
Mainline version of https://github.com/thesofproject/sof/pull/7564 

Link: https://github.com/thesofproject/sof/issues/7548

### CODE REVIEW

**Comment by kv2019i** [comment]
FYI @bardliao who worked on this area with the ES_BCLK feature.

File: src/drivers/intel/ssp/ssp.c [hunk 1]
```diff
@@ -975,6 +975,11 @@ static void ssp_early_start(struct dai *dai, int direction)
 
 	key = k_spin_lock(&dai->lock);
 
+	if (direction == DAI_DIR_CAPTURE) {
+		/* RX fifo must be cleared before start */
+		ssp_empty_rx_fifo(dai);
+	}
```
Comment (ujfalusi) [inline src/drivers/intel/ssp/ssp.c:981]: niptick:
```c
/* Before starting a capture the RX FIFO must be cleared */
if (direction == DAI_DIR_CAPTURE)
	ssp_empty_rx_fifo(dai);
```

File: src/drivers/intel/ssp/ssp.c [hunk 2]
```diff
@@ -87,9 +96,12 @@ static void ssp_empty_rx_fifo(struct dai *dai)
...
 		entries = SSCR3_RFL_VAL(ssp_read(dai, SSCR3));
 		dai_dbg(dai, "ssp_empty_rx_fifo(), before flushing, entries %d", entries);
-		for (i = 0; i < entries + 1; i++)
-			/* read to try empty fifo */
-			ssp_read(dai, SSDR);
+
+		if (read_ssdr) {
+			for (i = 0; i < entries + 1; i++)
+				/* read to try empty fifo */
+				ssp_read(dai, SSDR);
```
Comment (ujfalusi) [inline src/drivers/intel/ssp/ssp.c:103]: nitpick: I would move the comment up, before the for() loop and reword it to `Empty the FIFO by reading out the entries`

Comment (kv2019i) [reply src/drivers/intel/ssp/ssp.c:103]: Done

File: src/drivers/intel/ssp/ssp.c (hunk 1 above)
Comment (kv2019i) [reply src/drivers/intel/ssp/ssp.c:980]: Changed (although noting that in Zephyr, braces are mandatory here, trying to learn this style as this gets us rejects on Zephyr side all the time).

Comment (ujfalusi) [reply src/drivers/intel/ssp/ssp.c:980]: Well, if you have the comment inside of the if, then yes, the braces help, but I suggested to move it _outside_ of the if:
```c
/* Before starting a capture the RX FIFO must be cleared */
if (direction == DAI_DIR_CAPTURE)
	ssp_empty_rx_fifo(dai);
```
Does Zephyr mandates that braces are mandatory in all instance?

**Review by ujfalusi — APPROVED** [review]

File: src/drivers/intel/ssp/ssp.c [hunk 3]
```diff
@@ -74,9 +74,18 @@ static void ssp_empty_rx_fifo(struct dai *dai)
 	uint64_t sample_ticks = clock_ticks_per_sample(PLATFORM_DEFAULT_CLOCK,
 						       ssp->params.fsync_rate);
 	uint32_t retry = SSP_RX_FLUSH_RETRY_MAX;
+	bool direct_reads = ssp->state[DAI_DIR_CAPTURE] <= COMP_STATE_PREPARE;
```
Comment (btian1) [inline src/drivers/intel/ssp/ssp.c:77]: how about COMP_STATE_PAUSED? it is not active, and seems will not read with this state.

Comment (kv2019i) [reply src/drivers/intel/ssp/ssp.c:77]: @btian1 That's a good question. Earlier version of the patch broke PAUSE, so I did spend some time looking at this. Curiously, ssp_pause does not stop the hardware and it based on tests, this is not needed after pause. I can't fully explain why this is the case (and DMA is stopped on pause).

Comment (ujfalusi) [reply src/drivers/intel/ssp/ssp.c:77]: the other patch fixes the pause by yanking the FIFO before resume (which is a start operation), it is really strange to be honest, but w/o the FIFO drain patch I had broken pause/resume and I think @kv2019i had a working one on his (identical) setup.
